enigma-runtime-t/src/data/state.rs

```diff
@@ -54,28 +55,34 @@ impl IOInterface<EnclaveError, u8> for ContractState {
     }
 }
 
-impl DeltasInterface<EnclaveError, StatePatch> for ContractState {
-    fn apply_delta(&mut self, delta: &StatePatch) -> Result<(), EnclaveError> {
-        if delta.previous_hash != self.delta_hash {
-            return Err(SystemError(StateError { err: "Applying the delta".to_string() }));
+impl<'a> DeltasInterface<EnclaveError, EncryptedPatch, &'a StateKey> for ContractState {
+    fn apply_delta(&mut self, delta: EncryptedPatch, key: &'a StateKey) -> Result<(), EnclaveError> {
+        let delta_hash = delta.keccak256_patch();
+        let dec_delta = StatePatch::decrypt(delta.clone(), key)?;
+        if dec_delta.previous_hash != self.delta_hash {
+            return Err(SystemError(StateError { err: "Hashes don't match, Failed Applying the delta".to_string() }));
         }
-        json_patch::patch(&mut self.json, &delta.patch)?;
-        self.delta_hash = delta.sha256_patch()?;
-        self.delta_index = delta.index;
+        json_patch::patch(&mut self.json, &dec_delta.patch)?;
+        self.delta_hash = delta_hash;
+        self.delta_index = dec_delta.index;
         Ok(())
     }
 
-    fn generate_delta_and_update_state(old: &Self, new: &mut Self) -> Result<StatePatch, EnclaveError> {
-        new.delta_index = &old.delta_index+1;
-        let result = StatePatch{
+    fn generate_delta_and_update_state(old: &Self, new: &mut Self, key: &'a StateKey) -> Result<EncryptedPatch, EnclaveError> {
+        if old.delta_hash.is_zero() {
```

Comment:
Maybe it is better to extract the code of getting delta index from delta to a function, this will be cleaner. 
I think that the differentiation of the empty and non-empty delta this way is fine.